drivers/overlay/ov_network.go

```diff
@@ -1031,22 +1036,25 @@ func (n *network) releaseVxlanID() ([]uint32, error) {
 		}
 	}
 	var vnis []uint32
+	n.Lock()
 	for _, s := range n.subnets {
 		if n.driver.vxlanIdm != nil {
-			vni := n.vxlanID(s)
-			vnis = append(vnis, vni)
-			n.driver.vxlanIdm.Release(uint64(vni))
+			vnis = append(vnis, s.vni)
 		}
+		s.vni = 0
+	}
+	n.Unlock()
 
-		n.setVxlanID(s, 0)
+	for _, vni := range vnis {
+		n.driver.vxlanIdm.Release(uint64(vni))
 	}
 
 	return vnis, nil
 }
 
 func (n *network) obtainVxlanID(s *subnet) error {
 	//return if the subnet already has a vxlan id assigned
-	if s.vni != 0 {
+	if n.vxlanID(s) != 0 {
```

Comment:
Other places you replaced n.vxlanID(s) with s.vni or n.vni.   Here we are using n.vxlanID(s) instead of directly accessing with s.vni . Trying to understand why this change is required ?

Reply:
Also fair.   The revamped locking holds the network lock through network.join() and network.leave() operations.  As such, calling `n.vxlanID(s)` within those methods (or their sub-functions) would be a double lock.  That's the reasons for the changes above.

However, the changes above also *remove* the subnet locks and place subnet locking all within the domain of the network's lock.   As such, `network` methods need to acquire the network lock before accessing any of the subnet fields.  This driver calls obtainVxlanID() outside of the context of `network.join()` in several places.  Hence accessing the subnet members requires doing so through the network lock.

The lack of locking in obtainVxlanId() was, in principle, a race condition before that just wasn't addressed.